ASTImporter::Imported - VisitFunctionTemplateDecl assertion fix

balazske opened the pull request:
Fix for assertion
"Try to import an already imported Decl"
that occurs through calling
ASTNodeImporter::VisitFunctionTemplateDecl.

This fixes #311.

balazske commented:
It is possible to remove a part of function `ASTNodeImporter::VisitFunctionDecl` as written in #311 but this is left out now, the fix works without this and with this change too. The change can be done as part of the #300.

Xazax-hun commented:
https://github.com/Ericsson/clang/pull/300 is now merged. As per our discussion with @szepet , we should not remove the code snippet in question.

gerazo commented on lib/AST/ASTImporter.cpp:
```diff
@@ -4606,11 +4601,21 @@ Decl *ASTNodeImporter::VisitFunctionTemplateDecl(FunctionTemplateDecl *D) {
     }
   }
 
```
The code from upper here was moved below this point. Is it because the previous code should not search in the TO context after putting something into it?

Xazax-hun commented on lib/AST/ASTImporter.cpp:
```diff
@@ -4664,6 +4664,11 @@ Decl *ASTNodeImporter::VisitFunctionTemplateDecl(FunctionTemplateDecl *D) {
     return nullptr;
 
+  // Check recursive import.
+  Decl *AlreadyImported = Importer.GetAlreadyImportedOrNull(D);
```
I am wondering what should we do with this piece of code right now.
When we are importing the templated FunctionDecl above we will call import on the described template. This will result in a second call to importing the function template where the above import call will return the already imported function decl. After that we finish importing the template. After we imported the function decl and return from the function above we try to finish importing the same function template decl the second time. While having this check is definitely solving the problem I am wondering if there is a nicer solution for this cases in general. While I do agree it is good to fix this problem for now, maybe it worth a fixme to note that this is something that might have a better solution some day. @szepet  what do you think?

Xazax-hun reviewed (commented):
Overall looks good, but I think this might be a quick fix not the proper solution. Maybe it would be great to add a FIXME. Also, don't we need test for this?

Xazax-hun commented:
Hmm, I am still thinking if this version or the previous one is the better. I wonder what is the opinion of the others? 
I lean towards the previous one with a FIXME comment. Because in the future if someone rewrites the Importer with a new architecture (e.g. more context available during importing nodes), one could introduce a cleaner solution.

My intuition: the current solution will create nicer call stacks, but might result in more lookups. The previous one is a smaller modification that might be easier to understand, but the call stack and the importing logic is more convoluted that way.

balazske commented:
There are still some problems with the import: When importing the "templated" function lookup of already imported is not working correctly (somewhat like in the #312 case). This causes additional complexity with the new solution. Maybe the previous version of import is the better one, and import of `RecordDecl` and `ClassTemplateDecl` is done in similar way.

Xazax-hun commented:
I see, in this case, I am in favor of the previous solution. It looks like this is already a pattern in the importer, let's not break it for now.

balazske commented on lib/AST/ASTImporter.cpp:
```diff
@@ -4659,19 +4673,20 @@ Decl *ASTNodeImporter::VisitFunctionTemplateDecl(FunctionTemplateDecl *D) {
-      cast_or_null<FunctionDecl>(Importer.Import(D->getTemplatedDecl()));
-  if (!TemplatedFD)
-    return nullptr;
+  // Check recursive import.
```
This is the second solution that tries to avoid the "loop" at template import. Always import the "templated function" first. If the import is called from the templated function (this is indicated by the condition that `TemplatedFD` is non-null) import really the template. If the import is called not from the templated function (`TemplatedFD` should be null) import the templated function, this causes a recursive call with the previous condition, template is imported, and this call returns the already imported template.

But currently the first solution is the likely one.

balazske committed:
Fix for import of templated function from template decl

Additionally, import of templated function should find the
template if it already exists.

balazske commented:
Changes are updated, this should be the final version. (Not the most nice solution but should work.)

martong reviewed (approved):
LGTM! Nice work! Perhaps it can close #319 as well.

balazske commented:
`clang check-all` passes with this code

balazske commented:
Upstream revision: https://reviews.llvm.org/D49235
(Not fully related change, other related things are fixed and change in #394 is taken into account.)

balazske commented:
Backport from upstream: #439
